### PR TITLE
DATALAB-4500: Add freq_of_recommended_items_at_k test

### PR DESCRIPTION
**Pull request by christinabo**
Implementes a test to calculate the frequency with which items are being recommended within the given recommendations set.

### CODE REVIEW

File: reclist/metrics/standard_metrics.py [hunk 1]
```diff
@@ -127,4 +133,62 @@ def precision_at_k(y_preds, y_test, k=3):
...
+    def _calculate_hist(frequencies, bins):
+        """ Works out the counts of items in each bucket """
+        counts_per_bin = Counter([bisect_right(bins, item) - 1 for item in frequencies])
+        counts_per_bin_list = list(counts_per_bin.items())
+        empty_bins_indices = [ele for ele in np.arange(len(bins) - 1) if ele not in [
+            index for index, _ in counts_per_bin_list
+        ]]
+        counts_per_bin_list.extend([(index, 0) for index in empty_bins_indices])
+        counts_per_bin_sorted = sorted(counts_per_bin_list, key=lambda x: x[0], reverse=False)
+        return [y for _, y in counts_per_bin_sorted]
```
Comment (alepiscopo) [inline reclist/metrics/standard_metrics.py:149]: Nice!

File: reclist/metrics/standard_metrics.py [hunk 2]
```diff
@@ -127,4 +133,62 @@ def precision_at_k(y_preds, y_test, k=3):
...
+    frequencies = list(counts.values())
+
+    # fixed bin size
+    bins = np.arange(np.min(frequencies), np.max(frequencies) + bin_width, bin_width)
+    counts_per_bin_sorted = _calculate_hist(frequencies, bins)
+
+    # log bin size
+    log_bins = np.logspace(np.log10(bins[0]), np.log10(bins[-1]), len(bins))
+    log_bins = np.array([int(round(log_bin, 2)) for log_bin in log_bins])
+    counts_per_logbin_sorted = _calculate_hist(frequencies, log_bins)
```
Comment (alepiscopo) [inline reclist/metrics/standard_metrics.py:172]: Would it be worth to make using log optional?

File: reclist/metrics/standard_metrics.py [hunk 3]
```diff
@@ -127,4 +133,62 @@ def precision_at_k(y_preds, y_test, k=3):
...
+
+        # debug / visualization
+        ax1.bar(bins[:-1], counts_per_bin_sorted,
+                width=bin_width, align='edge')
+
+        ax1.set_ylabel('Number of items')
+        ax2.bar(log_bins[:-1], counts_per_logbin_sorted,
+                width=(log_bins[1:] - log_bins[:-1]), align='edge')
+        ax2.set_xscale('log', base=10)
+        ax2.set_xlabel('Frequency of being recommended')
```
Comment (alepiscopo) [inline reclist/metrics/standard_metrics.py]: What about `No. of times an item appear in a recs list`? We can reduce the fontsize, if needed.

File: reclist/metrics/standard_metrics.py [hunk 4]
```diff
@@ -127,4 +133,62 @@ def precision_at_k(y_preds, y_test, k=3):
...
+        'histogram_fixed': _format_results(bins, counts_per_bin_sorted),
+        'histogram_log': _format_results(log_bins, counts_per_logbin_sorted)
+    }
+
+    if debug:
+        f, (ax1, ax2) = plt.subplots(2, 1)
+
+        # debug / visualization
+        ax1.bar(bins[:-1], counts_per_bin_sorted,
+                width=bin_width, align='edge')
```
Comment (alepiscopo) [inline reclist/metrics/standard_metrics.py:184]: I wonder if using another visualisation approach may be better to interpret the data in this case (e.g. a scatterplot, where we could colour code genres, if we wanted to expand it), but this is good for the time being.

**Review by alepiscopo — APPROVED** [review]
This looks very good, great work, Christina!

File: reclist/metrics/standard_metrics.py (hunk 1 above)
Comment (christinabo) [reply reclist/metrics/standard_metrics.py:149]: That was painful to compute manually! I appreciated the existence of libraries 😅

File: reclist/metrics/standard_metrics.py (hunk 2 above)
Comment (christinabo) [reply reclist/metrics/standard_metrics.py:172]: We could but I think it makes sense for this particular metric to generate both as there is usually skewness in the results. I can change it if it's a problem in the future.

File: reclist/metrics/standard_metrics.py (hunk 3 above)
Comment (christinabo) [reply reclist/metrics/standard_metrics.py]: Good point, I updated it, thanks!

File: reclist/metrics/standard_metrics.py (hunk 4 above)
Comment (christinabo) [reply reclist/metrics/standard_metrics.py:184]: Yes, possibly - we have the data ready so we can experiment with different ones.

File: reclist/metrics/standard_metrics.py [hunk 5]
```diff
@@ -127,4 +133,62 @@ def precision_at_k(y_preds, y_test, k=3):
...
+
+        ax1.set_ylabel('Number of items')
+        ax2.bar(log_bins[:-1], counts_per_logbin_sorted,
+                width=(log_bins[1:] - log_bins[:-1]), align='edge')
+        ax2.set_xscale('log', base=10)
+        ax2.set_xlabel('No. of times an item appear in a recs list')
```
Comment (alepiscopo) [inline reclist/metrics/standard_metrics.py:190]: Sorry, my mistake – there's a typo in the label I suggested, it should be `an item appears`.

**Pull request merged**